### DISCUSSION TITLE
[2024-04-15] Library & Engine Team Update

### DISCUSSION THREAD

**Discussion by GregoryTravis** [enso-org/enso#9695] [Weekly Demo Day]
- Radosław: Refactoring `Enso_File` to be path based #9581
- Radosław: Improve handling of `File.parent` and `File.name` #9642
- Radosław: Make HTTP tests more robust by adding retries to the tests #9652
- Radosław: Integrate Cloud path resolver #9662
- Radosław: Writing Cloud files #9686
- Greg: Decimal remainder, pow, div #9566
- Greg: `Decimal` arithmetic benchmarks #9582
- Greg: Add range info to `Out_Of_Range` error #9586
- Greg: Remove pending tests for hash consistency. #9588
- Greg: Make the`%` consistent across both `Integer` sizes. #9589
- Greg: Implement `Runtime.Context.Dataflow_Stack_Trace` for dataflow errors thrown from Enso #9624
- Greg: Implement Runtime.Context.Dataflow_Stack_Trace for dataflow errors thrown from Enso #9625
- Greg: Tests for floating point mod #9636
- Greg: `Decimal.parse` and `.format` #9637
- Greg: `Decimal.abs`, `.negate`, `.signum` #9641
- Greg: Attach warning when converting from `Decimal` to `Integer` #9660
- Greg: Add `Decimal.min` and `.max` #9663
- Greg: Add `Decimal.round` #9672
- Greg: On MacOS, do not assume existence of `Contents/Home` in the JDK binary path #9684
- Greg: Add `Decimal.floor`, `.ceil`, and `.trunc` #9694
- Cassandra: April monthly recap video
- Cassandra: Data cleansing component design #9693
- Cassandra: Revamped blogs
- Cassandra: Finished and published Google Analytics blog
- Cassandra: Completed database query video - https://youtu.be/Qpk67SdUmQo
- Cassandra: Created component feedback spreadsheet and started review of Table components
- Ad: Add google analytics sample data #9571
- Ad: Wip/adr/add more table running functionality #9577
- Jaroslav: Don't let the non-Enso types float around the Enso interpreter! #9584
- Jaroslav: Removing useless methods #9590
- Jaroslav: Autoscoped constructors in a vector #9630
- Jaroslav: Accept small BigInteger values #9655
- Hubert: Replace JSON's Javascript implementation with Jackson one #9591
- Hubert: Fix expectations in lock manager test #9592
- Hubert: Manage analytics client to avoid leaks #9611
- Hubert: Improve perf of Graph.Scope.scopeFor hotspot #9620
- Hubert: Follow up #9611 #9629
- Hubert: Experiment in lock-free module invalidation #9639
- Hubert: Open Java modules for Snowflake setup #9664
- Pavel: Equality does not swallow errors #9560
- Pavel: Do not disable BackgroundCompilation in benchmarks #9646
- Pavel: Update to GraalVM 24.0.0 #9647
- Pavel: Export works without import #9683
- Pavel: Atom constructors can be private #9692
- Dmitry: GUI project does not receive suggestion database #9634
- Dmitry: Language Server API for AI #9679

**Comment by JaroslavTulach** [comment]
Status of
- #9410

Imagine you have two tables. One with columns `["a", "b", "e"]` and second with columns `["c", "b", "e"]`...

![two tables joined](https://github.com/enso-org/enso/assets/26887752/b1285b07-6ff1-4370-9913-30ece7d681c5)

After adding an `on` condition, the widget offers only `"b"` and `"e"`...

![on widget can access self and right](https://github.com/enso-org/enso/assets/26887752/1d5ebbd3-ea1b-46a4-9a97-e1e31c223470)

Why? Because the [Table.Widgets_Helper can access](https://github.com/enso-org/enso/pull/9410/files#r1525067756) values of arguments of the `join` function. If zeroth argument or argument named `right` is a `Table`, then the widgets can compare columns of `self` and this table and **offer only meaningful suggestions** - e.g. only columns that are present in both tables.

This is status as of https://github.com/enso-org/enso/pull/9410/commits/81aa2160c99a8a0ba3484f167a3c917e0190981a and it seems to be stable - e.g. you should be able to see the same on your computer after running `sbt:enso> runProjectManagerDistribution`
